src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs

```diff
@@ -206,31 +206,13 @@ public void AddConstant(string name, FormulaValue data)
         internal void AddUserDefinedFunction(string script, CultureInfo parseCulture = null, ReadOnlySymbolTable symbolTable = null, ReadOnlySymbolTable extraSymbolTable = null)
         {
             // Phase 1: Side affects are not allowed.
-            var options = new ParserOptions() 
-            { 
-                AllowsSideEffects = false, 
-                Culture = parseCulture ?? CultureInfo.InvariantCulture 
-            };
-            var sb = new StringBuilder();
-
-            UserDefinitions.ProcessUserDefinitions(script, options, out var userDefinitionResult);
-
-            if (userDefinitionResult.HasErrors)
-            {
-                sb.AppendLine("Something went wrong when parsing user defined functions.");
-
-                foreach (var error in userDefinitionResult.Errors)
-                {
-                    error.FormatCore(sb);
-                }
-
-                throw new InvalidOperationException(sb.ToString());
-            }
+            (var userDefinedFunctions, _) = UserDefinitions.Process(script, parseCulture);

```

Comment:
I'm wondering if parsing (processing) the UDFs shouldn't be the responsibility of the Engine, rather than the symbol table. Logically, symbol table is more a place to store what symbols exist, so it probably should only come into play after the UserDefinition script is parsed/validated, I think

Reply:
Maybe instead of AddUserDefinition taking a script, it should take an instance of UDF? Maybe even one that's already bound?

Reply:
Moved to the engine.